src/persist/src/s3.rs

```diff
@@ -46,6 +52,56 @@ pub struct S3BlobConfig {
     prefix: String,
 }
 
+// There is no simple way to hook into the S3 client to capture when its various timeouts
+// are hit. Instead, we pass along marker values that inform our [MetricsSleep] impl which
+// type of timeout was requested so it can substitute in a dynamic value set by config
+// from the caller.
+const OPERATION_TIMEOUT_MARKER: Duration = Duration::new(111, 1111);
+const OPERATION_ATTEMPT_TIMEOUT_MARKER: Duration = Duration::new(222, 2222);
+const CONNECT_TIMEOUT_MARKER: Duration = Duration::new(333, 3333);
+const READ_TIMEOUT_MARKER: Duration = Duration::new(444, 4444);
+
+#[derive(Debug)]
+struct MetricsSleep {
+    knobs: Box<dyn BlobKnobs>,
+    metrics: S3BlobMetrics,
+}
+
+impl AsyncSleep for MetricsSleep {
+    fn sleep(&self, duration: Duration) -> Sleep {
+        let (duration, metric) = match duration {
+            OPERATION_TIMEOUT_MARKER => (
+                self.knobs.operation_timeout(),
+                Some(self.metrics.operation_timeouts.clone()),
+            ),
+            OPERATION_ATTEMPT_TIMEOUT_MARKER => (
+                self.knobs.operation_attempt_timeout(),
+                Some(self.metrics.operation_attempt_timeouts.clone()),
+            ),
+            CONNECT_TIMEOUT_MARKER => (
+                self.knobs.connect_timeout(),
+                Some(self.metrics.connect_timeouts.clone()),
+            ),
+            READ_TIMEOUT_MARKER => (
+                self.knobs.read_timeout(),
+                Some(self.metrics.read_timeouts.clone()),
+            ),
+            duration => (duration, None),
```

Comment:
when might we hit this case? is unexpected and something we'd want to detect?

Reply:
the sleep impl is passed around to some deep internals in the SDK, and I believe it does get used internally for timeouts unrelated to the top-level client knobs

Reply:
👍